### PR TITLE
3.1 specify auth realm

### PR DESCRIPTION
**Pull request by craigtaverner**
changelog: Support the ability for client applications to specify which authorizing realm to use when logging in

### CODE REVIEW

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/ShiroAuthToken.java [hunk 1]
```diff
@@ -56,4 +56,10 @@ public String getScheme() throws InvalidAuthTokenException
     {
         return authToken;
     }
+
+    /** returns true if token map does not specify a realm, or if it specifies the requested realm */
+    public boolean supportsRealm( String realm )
+    {
+        return !authToken.containsKey( AuthToken.REALM_KEY ) || authToken.get( AuthToken.REALM_KEY ).equals( realm );
```
Comment (henriknyman) [inline enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/ShiroAuthToken.java]: Should we also support empty string as realm?

File: community/security/src/main/java/org/neo4j/server/security/auth/BasicAuthManager.java [hunk 2]
```diff
@@ -90,6 +95,13 @@ public void shutdown() throws Throwable
     }
 
     @Override
+    public boolean supports( Map<String,Object> authToken )
+    {
+        return (!authToken.containsKey( SCHEME_KEY ) || BASIC_SCHEME.equals( authToken.get( SCHEME_KEY ) )) &&
```
Comment (pontusmelke) [inline community/security/src/main/java/org/neo4j/server/security/auth/BasicAuthManager.java]: when do we not provide a `SCHEME_KEY`?

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/InternalFlatFileRealm.java [hunk 3]
```diff
@@ -192,7 +192,9 @@ public boolean supports( AuthenticationToken token )
         {
             if ( token instanceof ShiroAuthToken )
             {
-                return ((ShiroAuthToken) token).getScheme().equals( "basic" );
+                ShiroAuthToken shiroAuthToken = (ShiroAuthToken) token;
```
Comment (pontusmelke) [inline enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/InternalFlatFileRealm.java:197]: These casts looks a little like we could have a better design. Some kind of visitor maybe?

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/LdapRealm.java [hunk 4]
```diff
@@ -194,37 +198,40 @@ protected AuthorizationInfo queryForAuthorizationInfo( PrincipalCollection princ
...
+                        // TODO: Do a new LDAP search? But we need to cache the credentials for that...
+                        // Or we need the resulting failure message to the client to contain some status
+                        // so that the client can react by resending the auth token.
+                    }
+                    return authorizationInfo;
                 }
-                return authorizationInfo;
             }
         }
         return null;
```
Comment (pontusmelke) [inline enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/LdapRealm.java:237]: Do we have to use `null` as defaults here?

File: community/bolt/src/test/java/org/neo4j/bolt/security/auth/BasicAuthenticationTest.java [hunk 5]
```diff
@@ -220,6 +233,18 @@ public void shouldFailOnMalformedToken() throws Exception
                 .authenticate( map( "scheme", "basic", "principal", singletonList( "bob" ), "credentials", "secret" ) );
     }
 
+    private void mockManagerSupportsScheme(BasicAuthManager manager, String expectedScheme) {
+        doAnswer(new Answer<Boolean>(){
```
Comment (Mats-SX) [inline community/bolt/src/test/java/org/neo4j/bolt/security/auth/BasicAuthenticationTest.java]: This doesn't follow our Java code style (not enough padding whitespace).

File: community/bolt/src/main/java/org/neo4j/bolt/security/auth/BasicAuthentication.java [hunk 6]
```diff
@@ -54,7 +52,7 @@ public BasicAuthentication( AuthManager authManager, LogProvider logProvider )
     @Override
     public AuthenticationResult authenticate( Map<String,Object> authToken ) throws AuthenticationException
     {
-        if ( !SCHEME.equals( authToken.get( SCHEME_KEY ) ) )
+        if ( !authManager.supports( authToken ) )
         {
             throw new AuthenticationException( Status.Security.Unauthorized, "Missing username and password" );
```
Comment (fickludd) [inline community/bolt/src/main/java/org/neo4j/bolt/security/auth/BasicAuthentication.java]: Maybe the error message should be 'Invalid auth token' or similar?

**Pull request merged**